### PR TITLE
Rest API for car rental app

### PR DESCRIPTION
**Pull request by nico-bt**
Go to Readme: https://github.com/nico-bt/backend-car-rental-API/blob/development/README.md

### Live deploys to see UI and interaction:
- Frontend: https://frontend-car-rental-ivory.vercel.app/
- API: https://car-rental-api.adaptable.app/api/clients

(Obs: The free tier goes to sleep after a while, and the first load may take a few seconds.)

Pull Request Frontend:
https://github.com/nico-bt/frontend-car-rental/pull/1

### CODE REVIEW

File: src/car/dto/update-car.dto.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  IsEnum,
+  IsInt,
+  IsOptional,
+  IsPositive,
+  IsString,
+  Max,
+  Min,
+} from 'class-validator';
+
+export class UpdateCarDto {
```
Comment (venutti) [inline src/car/dto/update-car.dto.ts]: If the only difference between CreateCarDto and UpdateCarDto is the use of `@IsNotEmpty`/`@IsOptional`, i think that you can use a mapped-type instead https://docs.nestjs.com/openapi/mapped-types#partial
```suggestion
export class UpdateCarDto extends PartialType(CreateCarDto)  {}
```

File: src/car/dto/create-car.dto.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  marca: string;
+
+  @IsString()
+  @IsNotEmpty()
+  modelo: string;
+
+  @IsNotEmpty()
+  @IsInt()
+  @Min(2015)
+  @Max(2023)
```
Comment (venutti) [inline src/car/dto/create-car.dto.ts]: You can store these magic numbers (and the following ones) in a `constants.ts` file

```ts
export const MIN_YEAR = 2015;
export const MAX_YEAR = 2023;
```

```suggestion
  @Min(MIN_YEAR)
  @Max(MAX_YEAR)
```

File: src/car/car.service.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        marca,
+        modelo,
+        year,
+        km,
+        color,
+        ac,
+        pasajeros,
+        cambios,
+      },
+    });
```
Comment (venutti) [inline src/car/car.service.ts]: I think you could pass the createCarDto to the `data` property in the same way you do in the `update` method

```suggestion
    const newCar = await this.prismaService.car.create({
      data: createCarDto,
    });
```

File: src/car/dto/update-car.dto.ts (hunk 1 above)
Comment (nico-bt) [reply src/car/dto/update-car.dto.ts]: I didn't know about that feature, very practical.
I tested it locally and all the validations work as well.

File: src/car/dto/create-car.dto.ts (hunk 2 above)
Comment (nico-bt) [reply src/car/dto/create-car.dto.ts]: You are right, that is more clear, also i think actual year should be dynamic
```
const MIN_YEAR=2015
const ACTUAL_YEAR= new Date().getFullYear()
```

**Comment by FacundoCastillo12** [comment]
Hi, I took a quick look at it and here are some recommendations. One would be not to use Spanish and English together in the code, what I would recommend is to decide for one. Personally, I would recommend using English for everything, as you have done in much of the code.

As for the tests I think they are fantastic and must be tremendous work. This more than a recommendation would be a suggestion for the future, if you want to save time you could try to create all the tests as end-to-end and test each endpoint from the controller with supertest as if you were actually using the application. I saw you did a little bit of that, but it could be extended to everything. You could do it using SQLite as a database that runs only in test. That way you wouldn't have to test the way you did. While it can be tricky to create that testing environment, once it is set up in your application it can help you reduce testing time in the long run. 

Another would be to give some order to the folders. I do not say to follow an architecture because it can be complicated, but to divide it in different folders to begin with can help to leave the code more organized. You could create a folder for the controller, one for the service inside each module and inside those folders you could create another test folder to put the separate tests.

**Comment by nico-bt** [comment]
@FacundoCastillo12 gracias por el comment. 
Si, se me mezclaron los idiomas. Arranqué haciendo los modelos partiendo del .doc que hizo Fabricio con la consigna del ejercicio. Después seguí con todo en inglés y no cambié los modelos/forms...

I will look into your recommendation regarding end-to-end testing.
Thanks

**Pull request merged**